src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs

```diff
@@ -378,22 +244,20 @@ private bool Steal(out T result, bool take)
 #endif
 
             bool loop;
-            List<int> versionsList = new List<int>(); // save the lists version
+            List<int> versionsList = t_versionsList ?? (t_versionsList = new List<int>());
             do
             {
                 versionsList.Clear(); //clear the list from the previous iteration
                 loop = false;
 
-
-                ThreadLocalList currentList = _headList;
-                while (currentList != null)
+                ThreadLocalList currentList;
+                for (currentList = _headList; currentList != null; currentList = currentList._nextList)
                 {
                     versionsList.Add(currentList._version);
-                    if (currentList._head != null && TrySteal(currentList, out result, take))
+                    if (currentList.TrySteal(out result, take))
                     {
                         return true;
                     }
-                    currentList = currentList._nextList;
                 }
 
                 // verify versioning, if other items are added to this list since we last visit it, we should retry
```

Comment:
If I understand correctly, this versioning seems to cache failed attempts to steal (presumably because those lists were empty) and keeps trying as long as a list from any thread got a version update (transitioned from empty to nonempty). Is there a benefit to doing this? Why not just fail after the first attempt at stealing and ditch versioning altogether?